### PR TITLE
New SSH key flow

### LINKED ISSUE #62: Default to generating new ssh keys to use with loophole, add ssh-agent

### PR DESCRIPTION
**Pull request by denny241**
Closes #62

### CODE REVIEW

File: internal/app/loophole/loophole.go [hunk 1]
```diff
@@ -52,24 +57,51 @@ func setupCloseHandler() {
...
 
 func parsePublicKey(file string) (ssh.AuthMethod, ssh.PublicKey, error) {
-	key, err := ioutil.ReadFile(file)
+	privateKey, err := ioutil.ReadFile(file)
 
-	if err != nil {
+	var pathError *os.PathError
+	if errors.As(err, &pathError) { //if no keys are found, they are generated
+		var publicKey []byte
+		bitSize := 4096
```
Comment (0x7f) [inline internal/app/loophole/loophole.go:65]: Thumbs up for 4096 bits for RAS :) What about switching to `Ed25519` already? My private keys are all RSA still, but looks like Ed25519 is recommended these days? Does our loophole infrastructure even support it?

Comment (Morishiri) [reply internal/app/loophole/loophole.go:65]: Would need to check on gateway side

File: internal/app/loophole/loophole.go [hunk 2]
```diff
@@ -52,24 +57,51 @@ func setupCloseHandler() {
 }
 
 func parsePublicKey(file string) (ssh.AuthMethod, ssh.PublicKey, error) {
```
Comment (Morishiri) [inline internal/app/loophole/loophole.go:59]: I would like to make this whole ssh topic separate `package`, becuase this one grew too much.

File: internal/pkg/keys/keys.go [hunk 3]
```diff
@@ -1,42 +1,73 @@
...
 	"os"
 
-	"github.com/loophole/cli/internal/pkg/communication"
 	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
 	"golang.org/x/crypto/ssh/terminal"
 )
 
+//ParsePublicKey retrieves an ssh.AuthMethod and the related PublicKey
 func ParsePublicKey(terminalState *terminal.State, file string) (ssh.AuthMethod, ssh.PublicKey, error) {
```
Comment (Morishiri) [inline internal/pkg/keys/keys.go:20]: You missed changing the default file in `cmd/virtual-serve.go` line `25`. It should now be default get some files from `loophole` cache dir.

File: internal/pkg/keys/keys.go [hunk 4]
```diff
@@ -45,3 +76,83 @@ func ParsePublicKey(terminalState *terminal.State, file string) (ssh.AuthMethod,
...
+	signers, err := agentClient.Signers()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return signers[index], nil
+}
+
+//keySavedInSSHAgent goes through the identities saved in SSH Agent and looks for a specific key. If found, it returns true and the index, otherwise false and -1.
```
Comment (Morishiri) [inline internal/pkg/keys/keys.go:149]: Please split longer comments in two (or more)  lines so it's better readable

**Review by Morishiri — CHANGES_REQUESTED** [review]
Left 2 comments for you, but overall - great job :1st_place_medal:

**Review by Morishiri — APPROVED** [review]

**Pull request merged**